flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamRetractionRule.scala

```diff
@@ -0,0 +1,341 @@
+    * nodes need retraction from it. Currently, [[DataStreamOverAggregate]] and
+    * [[DataStreamGroupWindowAggregate]] need retraction from upstream nodes, besides, a
+    * needToRetract node also need retraction from it's upstream nodes.
+    */
+  class NeedToRetractProcessRule extends RelOptRule(
+    operand(
+      classOf[DataStreamRel], none()),
+    "NeedToRetractProcessRule") {
+
+    /**
+      * Return true if bottom RelNode does not contain needToRetract and top RelNode need
+      * retraction from bottom RelNode. Currently, operators which contain aggregations need
+      * retraction from upstream nodes, besides, a needToRetract node also needs retraction from
+      * it's upstream nodes.
+      */
+    def bottomNeedToRetract(topRel: RelNode, bottomRel: RelNode): Boolean = {
+      val bottomTraits = bottomRel.getTraitSet
+      if(!traitSetContainNeedToRetract(bottomTraits)){
+        topRel match {
+          case _: DataStreamGroupAggregate => true
+          case _: DataStreamGroupWindowAggregate => true
+          case _: DataStreamOverAggregate => true
+          case _ if traitSetContainNeedToRetract(topRel.getTraitSet) => true
+          case _ => false
+        }
+      } else {
+        false
+      }
+    }
+
+    /**
+      * Add needToRetract for the input RelNode
+      */
+    def addNeedToRetract(relNode: RelNode): RelNode = {
+      val traitSet = relNode.getTraitSet
+      var retractionTrait = traitSet.getTrait(RetractionTraitDef.INSTANCE)
+      if (null == retractionTrait) {
+        retractionTrait = new RetractionTrait(true, AccMode.Acc)
+      } else {
+        retractionTrait = new RetractionTrait(true, retractionTrait.getAccMode)
+      }
+
+      relNode.copy(traitSet.plus(retractionTrait), relNode.getInputs)
+    }
+
+    /**
+      * Returns a new topRel and a needTransform flag for a given topRel and bottomRels. The new
+      * topRel contains new bottomRels with needToRetract properly marked. The needTransform flag
+      * will be true if any transformation has been done.
+      *
+      * @param topRel The input top RelNode.
+      * @param bottomRels The input bottom RelNodes.
+      * @return A tuple holding a new top RelNode and a needTransform flag
+      */
+    def needToRetractProcess(
+        topRel: RelNode,
+        bottomRels: ListBuffer[RelNode])
+    : (RelNode, Boolean) = {
+
+      var needTransform = false
+      var i = 0
+      while(i < bottomRels.size) {
+        val bottomRel = bottomRels(i)
+        if(bottomNeedToRetract(topRel, bottomRel)) {
+          needTransform = true
+          bottomRels(i) = addNeedToRetract(bottomRel)
+        }
+        i += 1
+      }
+      val newTopRel = topRel.copy(topRel.getTraitSet, bottomRels.asJava)
+      (newTopRel, needTransform)
+    }
+
+    override def onMatch(call: RelOptRuleCall): Unit = {
+      val topRel = call.rel(0).asInstanceOf[DataStreamRel]
```

Comment:
usually `top` denotes the top node of a tree and `bottom` the leave nodes of a tree. I think `parent` and `children` / `child` would be more appropriate terms here.